pkg/controller/atlasproject/maintenancewindow.go

```diff
@@ -0,0 +1,164 @@
+package atlasproject
+
+import (
+	"context"
+	"errors"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/project"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+)
+
+// ensureMaintenanceWindow ensures that the state of the Atlas Maintenance Window matches the
+// state of the Maintenance Window specified in the project CR. If a Maintenance Window exists
+// in Atlas but is not specified in the CR, it is deleted.
+func ensureMaintenanceWindow(ctx *workflow.Context, projectID string, atlasProject *mdbv1.AtlasProject) workflow.Result {
+	windowSpec := atlasProject.Spec.ProjectMaintenanceWindow
+	if err := validateMaintenanceWindow(windowSpec); err != nil {
+		return workflow.Terminate(workflow.ProjectWindowInvalid, err.Error())
+	}
+
+	if isEmptyWindow(windowSpec) {
+		ctx.Log.Debugw("Window empty or undefined, deleting in Atlas")
+		if result := deleteInAtlas(ctx.Client, projectID); !result.IsOk() {
+			return result
+		}
+		return workflow.OK()
+	}
+
+	if windowSpecified(windowSpec) {
+		ctx.Log.Debugw("Checking if window needs update")
+		windowInAtlas, result := getInAtlas(ctx.Client, projectID)
+		if !result.IsOk() {
+			return result
+		}
+
+		if windowInAtlas.DayOfWeek == 0 || windowInAtlas.HourOfDay == nil ||
+			*windowInAtlas.HourOfDay != windowSpec.HourOfDay || windowInAtlas.DayOfWeek != windowSpec.DayOfWeek {
+			ctx.Log.Debugw("Creating or updating window")
+			// We set startASAP to false because the operator takes care of calling the API a second time if both
+			// startASAP and the new maintenance timeslots are defined
+			if result := createOrUpdateInAtlas(ctx.Client, projectID, windowSpec.WithStartASAP(false)); !result.IsOk() {
+				return result
+			}
+		} else if *windowInAtlas.AutoDeferOnceEnabled != windowSpec.AutoDefer {
+			// If autoDefer flag is different in Atlas, and we haven't updated the window previously, we toggle the flag
+			ctx.Log.Debugw("Toggling autoDefer")
+			if result := toggleAutoDeferInAtlas(ctx.Client, projectID); !result.IsOk() {
+				return result
+			}
+		}
+	}
+
+	if windowSpec.StartASAP {
+		ctx.Log.Debugw("Starting maintenance ASAP")
+		// To avoid any unexpected behavior, we send a request to the API containing only the StartASAP flag,
+		// although the API should ignore other fields in that case
+		if result := createOrUpdateInAtlas(ctx.Client, projectID, project.NewMaintenanceWindow().WithStartASAP(true)); !result.IsOk() {
+			return result
+		}
+		// Nothing else should be done after sending a StartASAP request
+		return workflow.OK()
+	}
+
+	if windowSpec.Defer {
```

Comment:
Shouldn't `startASAP` and `Defer` be mutually exclusive? What happens if I specify both?

Reply:
The function `validateMaintenanceWindow` should take care of that, should I also do the check in the Reconcile loop ?